File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.cc
#include <bits/stdc++.h>
using namespace std;

vector<int> adj[1003];
int n, m, visited[1003];

int main() {
    cin >> n >> m;
    while(m--) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    int num = 0;
    for(int i = 1; i <= n; i++) {
        if(visited[i]) continue;
        num++;
        queue<int> q;
        q.push(i);
        visited[i] = 1;
        while(q.size()) {
            int cur = q.front();
            q.pop();
            for(auto nxt : adj[cur]) {
                if(visited[nxt]) continue;
                q.push(nxt);
                visited[nxt] = 1;
            }
        }
    }
    cout << num;
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        //그래프 초기화
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        //visited 초기화
        visited = new boolean[n + 1];

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        int connectedComponents = 0;

        //모든 정점에 대해 BFS 수행
        for (int i = 1; i <= n; i++) {
            if (!visited[i]) {
                Queue<Integer> queue = new LinkedList<>();
                queue.add(i);
                while (!queue.isEmpty()) {
                    int cur = queue.poll();
                    for (int nxt : graph.get(cur)) {
                        if (visited[nxt]) continue;
                        visited[nxt] = true;
                        queue.add(nxt);
                    }
                }
                connectedComponents++;
            }
        }
        System.out.println(connectedComponents);

    }
}